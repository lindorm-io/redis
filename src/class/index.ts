export * from "./RedisClient";
export * from "./RedisInMemoryClient";
