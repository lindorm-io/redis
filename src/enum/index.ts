export * from "./redis-connection-type";
