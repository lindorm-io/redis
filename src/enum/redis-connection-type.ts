export enum RedisConnectionType {
  MEMORY = "MEMORY",
  CACHE = "CACHE",
}
