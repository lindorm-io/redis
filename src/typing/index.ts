export * from "./cache";
export * from "./redis-client";
export * from "./redis-connection";
