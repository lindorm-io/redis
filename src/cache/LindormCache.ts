import { CacheBase } from "./CacheBase";
import { CacheError, EntityNotFoundError } from "../error";
import { EntityCreationError, ILindormEntity, EntityAttributes } from "@lindorm-io/entity";
import { ILindormCache } from "../typing";
import { filter as _filter } from "lodash";

export abstract class LindormCache<Interface extends EntityAttributes, Entity extends ILindormEntity<Interface>>
  extends CacheBase
  implements ILindormCache<Interface, Entity>
{
  protected abstract createEntity(data: Interface): Entity;

  public async create(entity: Entity, expiresInSeconds?: number): Promise<Entity> {
    await entity.schemaValidation();

    const start = Date.now();
    const json = entity.toJSON();
    const key = `${this.prefix}::${entity.getKey()}`;

    try {
      entity.create();
    } catch (err) {
      if (!(err instanceof EntityCreationError)) {
        throw err;
      }
    }

    const result = await this.client.set(key, json, expiresInSeconds || this.expiresInSeconds);
    const success = result === "OK";

    this.logger.debug("create", {
      payload: json,
      result: { success },
      time: Date.now() - start,
    });

    if (!success) {
      throw new CacheError("Unable to set entity in cache", {
        debug: { key, json, result },
      });
    }

    return entity;
  }

  public async update(entity: Entity, expiresInSeconds?: number): Promise<Entity> {
    await entity.schemaValidation();

    const start = Date.now();
    const json = entity.toJSON();
    const key = `${this.prefix}::${entity.getKey()}`;

    const result = await this.client.set(key, json, expiresInSeconds);
    const success = result === "OK";

    this.logger.debug("update", {
      payload: json,
      result: { success },
      time: Date.now() - start,
    });

    if (!success) {
      throw new CacheError("Unable to set entity in cache", {
        debug: { key, json, result },
      });
    }

    return entity;
  }

  public async find(key: string): Promise<Entity> {
    const start = Date.now();

    const prefixKey = `${this.prefix}::${key}`;
    const result: unknown = await this.client.get(prefixKey);

    this.logger.debug("find", {
      key: prefixKey,
      result: { success: !!result },
      time: Date.now() - start,
    });

    if (!result) {
      throw new EntityNotFoundError("Unable to find entity in cache", {
        debug: { prefixKey, result },
      });
    }

    return this.createEntity(result as Interface);
  }

  public async findMany(filter: Partial<Interface>): Promise<Array<Entity>> {
    return _filter(await this.findAll(), filter) as Array<Entity>;
  }

  public async findAll(): Promise<Array<Entity>> {
    const start = Date.now();

    const pattern = `${this.prefix}::*`;
    const result = (await this.client.getAll(pattern)) as Array<Interface>;
    const data = [];

    for (const item of Object.values(result)) {
      data.push(this.createEntity(item));
    }

    this.logger.debug("findAll", {
      pattern,
      result: { success: !!result },
      time: Date.now() - start,
    });

    return data;
  }

  public async remove(entity: Entity): Promise<void> {
    const start = Date.now();

    const key = entity.getKey();
    const prefixKey = `${this.prefix}::${key}`;

    const deletedRows = await this.client.del(prefixKey);

    this.logger.debug("remove", {
      key: prefixKey,
      result: { success: !!deletedRows },
      time: Date.now() - start,
    });

    if (deletedRows === 0) {
      throw new CacheError("Unable to delete entity from cache", {
        debug: { prefixKey, deletedRows },
      });
    }
  }
}
