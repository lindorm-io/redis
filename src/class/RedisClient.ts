import redis, { ClientOpts } from "redis";
import { IRedisClient, TPromise } from "../typing";
import { isNumber } from "lodash";
import { parseBlob, stringifyBlob } from "../util";
import { promisify } from "util";

export interface IAsyncClient {
  del: TPromise<number>;
  get: TPromise<string>;
  keys: TPromise<string>;
  quit: TPromise<string>;
  set: TPromise<string>;
  setEx: TPromise<string>;
}

export class RedisClient implements IRedisClient {
  private _client: IAsyncClient;
  private _clientOptions: ClientOpts;

  constructor(options: ClientOpts) {
    this._clientOptions = options;
  }

  public async connect(): Promise<void> {
    const client: redis.RedisClient = redis.createClient(this._clientOptions);

    this._client = {
      del: promisify(client.del).bind(client),
      get: promisify(client.get).bind(client),
      keys: promisify(client.keys).bind(client),
      quit: promisify(client.quit).bind(client),
      set: promisify(client.set).bind(client),
      setEx: promisify(client.setex).bind(client),
    };
  }

  public async quit(): Promise<string> {
    return this._client.quit();
  }

  public async set(key: string, value: Record<string, any>, expiresInSeconds?: number): Promise<string> {
    let result: string;

    if (isNumber(expiresInSeconds)) {
      result = await this._client.setEx(key, expiresInSeconds, stringifyBlob(value));
    } else {
      result = await this._client.set(key, stringifyBlob(value));
    }

    return result;
  }

  public async get(key: string): Promise<Record<string, any>> {
    return parseBlob(await this._client.get(key));
  }

  public async getAll(pattern: string): Promise<Array<Record<string, any>>> {
    const keys = await this._client.keys(pattern);
    const array: Array<Record<string, any>> = [];

    for (const key of keys) {
      array.push(await this.get(key));
    }

    return array;
  }

  public async del(key: string): Promise<number> {
    return this._client.del(key);
  }
}
