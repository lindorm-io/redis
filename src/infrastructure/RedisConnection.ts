import { IRedisConnection, IRedisConnectionOptions, TRedisClient } from "../typing";
import { RedisConnectionBase } from "./RedisConnectionBase";
import { RedisConnectionCache } from "./RedisConnectionCache";
import { RedisConnectionMemory } from "./RedisConnectionMemory";
import { RedisConnectionType } from "../enum";

export class RedisConnection implements IRedisConnection {
  private connection: RedisConnectionBase;

  constructor(options: IRedisConnectionOptions) {
    switch (options.type) {
      case RedisConnectionType.CACHE:
        this.connection = new RedisConnectionCache(options);
        break;

      case RedisConnectionType.MEMORY:
        this.connection = new RedisConnectionMemory(options);
        break;

      default:
        throw new Error(`Unknown connection type: [ ${options.type} ]`);
    }
  }

  public async connect(): Promise<void> {
    return this.connection.connect();
  }

  public async disconnect(): Promise<void> {
    return this.connection.disconnect();
  }

  public getClient(): TRedisClient {
    return this.connection.getClient();
  }
}
