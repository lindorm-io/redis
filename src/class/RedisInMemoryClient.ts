import { IRedisClientOptions } from "./RedisClient";
import { TObject } from "@lindorm-io/core";
import { includes } from "lodash";
import { parseBlob, stringifyBlob } from "../util";

export class RedisInMemoryClient {
  public store: TObject<any>;

  constructor(options: IRedisClientOptions) {
    this.store = {};
  }

  public async set(key: string, value: TObject<any>, expiresInSeconds?: number): Promise<string> {
    this.store[key] = { blob: stringifyBlob(value), expiresInSeconds };

    return "OK";
  }

  public async get(key: string): Promise<TObject<any>> {
    const data = this.store[key];

    return data?.blob ? parseBlob(data.blob) : undefined;
  }

  public async getAll(pattern: string): Promise<Array<TObject<any>>> {
    const array: Array<TObject<any>> = [];
    const inMemoryPattern = pattern.replace(/\*/g, "");

    for (const key of Object.keys(this.store)) {
      if (!includes(key, inMemoryPattern)) continue;
      array.push(await this.get(key));
    }

    return array;
  }

  public async del(key: string): Promise<number> {
    this.store[key] = undefined;
    return 1;
  }
}
