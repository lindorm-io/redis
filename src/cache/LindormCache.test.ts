import Joi from "joi";
import MockDate from "mockdate";
import { LindormCache } from "./LindormCache";
import {
  EntityAttributes,
  EntityCreationError,
  EntityOptions,
  ILindormEntity,
  JOI_ENTITY_BASE,
  LindormEntity,
} from "@lindorm-io/entity";
import { RedisConnection } from "../infrastructure";
import { RedisConnectionType } from "../enum";
import { CacheOptions } from "../typing";
import { logger } from "../test";
import { RedisClient } from "../class";

MockDate.set("2020-01-01T08:00:00.000Z");

interface ITestEntityAttributes extends EntityAttributes {
  name: string;
  hasCreatedFunctionBeenCalled: boolean;
}

interface ITestEntityOptions extends EntityOptions {
  name: string;
  hasCreatedFunctionBeenCalled?: boolean;
}

interface ITestEntity extends ILindormEntity<ITestEntityAttributes> {}

const schema = Joi.object({
  ...JOI_ENTITY_BASE,
  name: Joi.string().required(),
  hasCreatedFunctionBeenCalled: Joi.boolean().required(),
});

class TestEntity extends LindormEntity<ITestEntityAttributes> implements ITestEntity {
  public name: string;
  public hasCreatedFunctionBeenCalled: boolean;

  constructor(options: ITestEntityOptions) {
    super(options);
    this.name = options.name;
    this.hasCreatedFunctionBeenCalled = options.hasCreatedFunctionBeenCalled || false;
  }

  create() {
    if (this.hasCreatedFunctionBeenCalled) {
      throw new EntityCreationError("TestEntity");
    }
    this.hasCreatedFunctionBeenCalled = true;
  }

  getKey() {
    return this.id;
  }

  async schemaValidation() {
    await schema.validateAsync(this.toJSON());
  }

  toJSON() {
    return {
      ...this.defaultJSON(),
      name: this.name,
      hasCreatedFunctionBeenCalled: this.hasCreatedFunctionBeenCalled,
    };
  }
}

class TestCache extends LindormCache<ITestEntityAttributes, TestEntity> {
  constructor(options: CacheOptions) {
    super({
      ...options,
      entityName: "mock",
    });
  }

  protected createEntity(data: ITestEntityAttributes): TestEntity {
    return new TestEntity(data);
  }
}

describe("LindormCache", () => {
  let inMemoryCache: Record<string, any>;
  let redis: RedisConnection;
  let client: RedisClient;
  let cache: TestCache;
  let entity: TestEntity;

  beforeEach(async () => {
    inMemoryCache = {};

    redis = new RedisConnection({
      type: RedisConnectionType.MEMORY,
      port: 1,
      inMemoryCache,
    });

    await redis.connect();
    client = redis.client();

    cache = new TestCache({ logger, client });
    entity = new TestEntity({
      id: "10192b05-6d39-4a2d-907a-32318fda8cb9",
      name: "name",
    });
  });

  afterEach(async () => {
    await client.quit();
    jest.clearAllMocks();
  });

  test("should create entity", async () => {
    await expect(cache.create(entity)).resolves.toMatchSnapshot();

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should create entity with expiry", async () => {
    await expect(cache.create(entity, 999)).resolves.toMatchSnapshot();

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should create entity with expiry", async () => {
    cache = new TestCache({ client, expiresInSeconds: 100, logger });

    await expect(cache.create(entity)).resolves.toMatchSnapshot();

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should find entity", async () => {
    await cache.create(entity);

    await expect(cache.find(entity.id)).resolves.toMatchSnapshot();
  });

  test("should update entity", async () => {
    await cache.create(entity);

    entity.name = "new name";

    await expect(cache.update(entity)).resolves.toMatchSnapshot();
  });

  test("should find all entities", async () => {
    await cache.create(
      new TestEntity({
        id: "5a8c3582-0a5f-4af8-b393-ac95d65284ee",
        name: "one",
      }),
    );
    await cache.create(
      new TestEntity({
        id: "a9afb688-b9c0-42ce-9fe2-1e2f76fee036",
        name: "two",
      }),
    );

    await expect(cache.findAll()).resolves.toMatchSnapshot();
  });

  test("should find many entities", async () => {
    await cache.create(
      new TestEntity({
        id: "0868b84c-ba82-49b2-968e-eba9c02286dd",
        name: "one",
      }),
    );
    await cache.create(
      new TestEntity({
        id: "712606fb-f371-44fe-b6b0-0aa6807950d8",
        name: "two",
      }),
    );
    await cache.create(
      new TestEntity({
        id: "8e84c5b7-f1fa-4c73-a342-2da513ffa1df",
        name: "two",
      }),
    );

    await expect(cache.findMany({ name: "two" })).resolves.toMatchSnapshot();
  });

  test("should remove entity", async () => {
    await cache.create(entity);

    await expect(cache.remove(entity)).resolves.toBeUndefined();

    expect(inMemoryCache).toMatchSnapshot();
  });
});
