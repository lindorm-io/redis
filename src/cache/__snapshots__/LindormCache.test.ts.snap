// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LindormCache should create entity 1`] = `
Object {
  "created": 2020-01-01T08:00:00.000Z,
  "events": Array [],
  "hasCreatedFunctionBeenCalled": true,
  "id": "10192b05-6d39-4a2d-907a-32318fda8cb9",
  "name": "name",
  "updated": 2020-01-01T08:00:00.000Z,
  "version": 0,
}
`;

exports[`LindormCache should create entity 2`] = `
Object {
  "mock::10192b05-6d39-4a2d-907a-32318fda8cb9": Object {
    "blob": "{\\"json\\":{\\"id\\":\\"10192b05-6d39-4a2d-907a-32318fda8cb9\\",\\"created\\":\\"2020-01-01T08:00:00.000Z\\",\\"events\\":[],\\"updated\\":\\"2020-01-01T08:00:00.000Z\\",\\"version\\":\\"0\\",\\"name\\":\\"name\\",\\"hasCreatedFunctionBeenCalled\\":\\"false\\"},\\"meta\\":{\\"id\\":\\"string\\",\\"created\\":\\"date\\",\\"events\\":[],\\"updated\\":\\"date\\",\\"version\\":\\"number\\",\\"name\\":\\"string\\",\\"hasCreatedFunctionBeenCalled\\":\\"boolean\\"}}",
    "expiresInSeconds": undefined,
  },
}
`;

exports[`LindormCache should create entity with expiry 1`] = `
Object {
  "created": 2020-01-01T08:00:00.000Z,
  "events": Array [],
  "hasCreatedFunctionBeenCalled": true,
  "id": "10192b05-6d39-4a2d-907a-32318fda8cb9",
  "name": "name",
  "updated": 2020-01-01T08:00:00.000Z,
  "version": 0,
}
`;

exports[`LindormCache should create entity with expiry 2`] = `
Object {
  "mock::10192b05-6d39-4a2d-907a-32318fda8cb9": Object {
    "blob": "{\\"json\\":{\\"id\\":\\"10192b05-6d39-4a2d-907a-32318fda8cb9\\",\\"created\\":\\"2020-01-01T08:00:00.000Z\\",\\"events\\":[],\\"updated\\":\\"2020-01-01T08:00:00.000Z\\",\\"version\\":\\"0\\",\\"name\\":\\"name\\",\\"hasCreatedFunctionBeenCalled\\":\\"false\\"},\\"meta\\":{\\"id\\":\\"string\\",\\"created\\":\\"date\\",\\"events\\":[],\\"updated\\":\\"date\\",\\"version\\":\\"number\\",\\"name\\":\\"string\\",\\"hasCreatedFunctionBeenCalled\\":\\"boolean\\"}}",
    "expiresInSeconds": 999,
  },
}
`;

exports[`LindormCache should create entity with expiry 3`] = `
Object {
  "created": 2020-01-01T08:00:00.000Z,
  "events": Array [],
  "hasCreatedFunctionBeenCalled": true,
  "id": "10192b05-6d39-4a2d-907a-32318fda8cb9",
  "name": "name",
  "updated": 2020-01-01T08:00:00.000Z,
  "version": 0,
}
`;

exports[`LindormCache should create entity with expiry 4`] = `
Object {
  "mock::10192b05-6d39-4a2d-907a-32318fda8cb9": Object {
    "blob": "{\\"json\\":{\\"id\\":\\"10192b05-6d39-4a2d-907a-32318fda8cb9\\",\\"created\\":\\"2020-01-01T08:00:00.000Z\\",\\"events\\":[],\\"updated\\":\\"2020-01-01T08:00:00.000Z\\",\\"version\\":\\"0\\",\\"name\\":\\"name\\",\\"hasCreatedFunctionBeenCalled\\":\\"false\\"},\\"meta\\":{\\"id\\":\\"string\\",\\"created\\":\\"date\\",\\"events\\":[],\\"updated\\":\\"date\\",\\"version\\":\\"number\\",\\"name\\":\\"string\\",\\"hasCreatedFunctionBeenCalled\\":\\"boolean\\"}}",
    "expiresInSeconds": 100,
  },
}
`;

exports[`LindormCache should find all entities 1`] = `
Array [
  Object {
    "created": 2020-01-01T08:00:00.000Z,
    "events": Array [],
    "hasCreatedFunctionBeenCalled": false,
    "id": "5a8c3582-0a5f-4af8-b393-ac95d65284ee",
    "name": "one",
    "updated": 2020-01-01T08:00:00.000Z,
    "version": 0,
  },
  Object {
    "created": 2020-01-01T08:00:00.000Z,
    "events": Array [],
    "hasCreatedFunctionBeenCalled": false,
    "id": "a9afb688-b9c0-42ce-9fe2-1e2f76fee036",
    "name": "two",
    "updated": 2020-01-01T08:00:00.000Z,
    "version": 0,
  },
]
`;

exports[`LindormCache should find entity 1`] = `
Object {
  "created": 2020-01-01T08:00:00.000Z,
  "events": Array [],
  "hasCreatedFunctionBeenCalled": false,
  "id": "10192b05-6d39-4a2d-907a-32318fda8cb9",
  "name": "name",
  "updated": 2020-01-01T08:00:00.000Z,
  "version": 0,
}
`;

exports[`LindormCache should find many entities 1`] = `
Array [
  Object {
    "created": 2020-01-01T08:00:00.000Z,
    "events": Array [],
    "hasCreatedFunctionBeenCalled": false,
    "id": "712606fb-f371-44fe-b6b0-0aa6807950d8",
    "name": "two",
    "updated": 2020-01-01T08:00:00.000Z,
    "version": 0,
  },
  Object {
    "created": 2020-01-01T08:00:00.000Z,
    "events": Array [],
    "hasCreatedFunctionBeenCalled": false,
    "id": "8e84c5b7-f1fa-4c73-a342-2da513ffa1df",
    "name": "two",
    "updated": 2020-01-01T08:00:00.000Z,
    "version": 0,
  },
]
`;

exports[`LindormCache should remove entity 1`] = `
Object {
  "mock::10192b05-6d39-4a2d-907a-32318fda8cb9": undefined,
}
`;

exports[`LindormCache should update entity 1`] = `
Object {
  "created": 2020-01-01T08:00:00.000Z,
  "events": Array [],
  "hasCreatedFunctionBeenCalled": true,
  "id": "10192b05-6d39-4a2d-907a-32318fda8cb9",
  "name": "new name",
  "updated": 2020-01-01T08:00:00.000Z,
  "version": 0,
}
`;
